buildSrc/src/main/java/com/google/firebase/gradle/plugins/semver/ApiDiffer.kt

```diff
@@ -0,0 +1,182 @@
+      deltaViolations.forEach { outputString += "[${it.versionDelta}] ${it.description}\n" }
+      throw GradleException(outputString)
+    }
+  }
+
+  private fun readApi(aarPath: File, destFile: String): Map<String, ClassInfo> {
+    UnzipAar.unzip(aarPath, destFile)
+    val classes: Map<String, ClassNode> = readClassNodes(Paths.get("${destFile}/classes.jar"))
+    var classInfoDict = mutableMapOf<String, ClassInfo>()
+    for (classNodeInfo in classes) {
+      classInfoDict[classNodeInfo.key] = ClassInfo(classNodeInfo.value, classes)
+    }
+    for (classNodeInfo in classes) {
+      classInfoDict[classNodeInfo.key] = ClassInfo(classNodeInfo.value, classes)
+    }
+    return classInfoDict
+  }
+
+  /** Returns true if the class is local or anonymous. */
+  private fun isLocalOrAnonymous(classNode: ClassNode): Boolean {
+    // JVMS 4.7.7 says a class has an EnclosingMethod attribute iff it is local or anonymous.
+    // ASM sets the "enclosing class" only if an EnclosingMethod attribute is present, so this
+    // this test will not include named inner classes even though they have enclosing classes.
+    return classNode.outerClass != null
+  }
+
+  private fun getUnqualifiedClassname(classNodeName: String): String {
+    // Class names may be "/" or "." separated depending on context. Normalize to "/"
+    val normalizedPath = classNodeName.replace(".", "/")
+    val withoutPackage = normalizedPath.substring(normalizedPath.lastIndexOf('/') + 1)
+    return withoutPackage.substring(withoutPackage.lastIndexOf('$') + 1)
+  }
+
+  fun readClassNodes(jar: Path): Map<String, ClassNode> {
+    val classes: MutableMap<String, ClassNode> = LinkedHashMap()
+    val inputStream = Files.newInputStream(jar)
+    val jis = JarInputStream(inputStream)
+    var je: JarEntry? = null
+    while (true) {
+      je = jis.nextJarEntry
+      if (je == null) {
+        break
+      }
+      if (!je.name.endsWith(CLASS_EXTENSION)) {
+        continue
+      }
+      val expectedName = je.name.substring(0, je.name.length - CLASS_EXTENSION.length)
+      val classNode: ClassNode = ClassNode()
+
+      ClassReader(jis).accept(classNode, ClassReader.SKIP_DEBUG or ClassReader.SKIP_FRAMES)
+
+      if (!classNode.name.equals(expectedName)) {
+        logger.error(
+          "Classnode doesn't match expected name ${classNode.name}. This is not a valid jar."
+        )
+        continue
+      }
+      // Skip classes that appear to be obfuscated.
+      if (UtilityClass.isObfuscatedSymbol(getUnqualifiedClassname(classNode.name))) {
+        continue
+      }
+      // Skip local and anonymous classes.
+      if (isLocalOrAnonymous(classNode)) {
+        continue
+      }
+      // Skip private nested classes
+      if (!classes.containsKey(classNode.name)) {
+        classes.put(classNode.name, classNode)
+      } else {
+        println("Duplicate class seen: " + classNode.name)
```

Comment:
Using the tasks logger may be better than these printlines

```kotlin
project.logger.info("Duplicate class seen: ${classNode.name}")
```

Reply:
+1